mautic/mautic
# Static segment members filter

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [n]
| New feature/enhancement? (use the a.x branch)      | [y]
| Deprecations?                          | [y]
| BC breaks? (use the c.x branch)        | [n]
| Automated tests included?              | [y] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | 



#### Description:

This PR is adding a new filter *Segment Membership - Static* that behaves pretty much like the existing *Segment Membership* filter with one difference:

The new filter use the already calculated segment members instead of calculating all the segment members for the child segments again.

The new filter is basically the Campaign Membership filter but it uses the segment membership table instead of the campaign membership table.

This new filter can take a segment with the segment membership filter that doesn't finish in days and finish it in 250ms according to one such segment that we were testing with.

The filter is disabled by default as an "experimental" feature. You have to add `'show_leadlist_static_filter' => true` to your config file to enable it. If it will prove to be accepted by our selected users, it will be enabled to everyone.

While testing all the features related to the `leadlist` segment filter, I found that the `CircularDependencyValidator` doesn't work, so I refactored it so it is using https://github.com/mautic/mautic/pull/11276 instead of fixing the code that:
1. wasn't working
2. worked only for 1 or 2 layers deep

Pros:
1. It will generate smaller, way more performant queries.
4. The work is not duplicated by building the child segments and building the child segments as part of the parent segment again.
5. Some segments with the segment membership filter generate so enormous queries that they will never finish. The new filter should let such filters finish.
6. Using this new filter will be cheaper to run and you may save some $$$ on infrastructure.

Cons:
1. It can happen that some segment members are in the child segment but shouldn't be because they weren't removed yet. Those members will appear in the parent segment too. With the original filter, all the filters were evaluated on the parent segment build and therefore the contacts were added or removed correctly.
2. In short, we can say that the original segment was slightly more accurate.
4. We'll have to teach our users which filter to use when.

#### Steps to test this PR:

Don't forget to enable the feature flag `show_leadlist_static_filter` to see the new filter.


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Test the new *Segment Membership - Static* that it works the same as the original *Segment Membership* filter
3. Test that if you add a segment to the *Segment Membership - Static* filter in a new segment then the dependent segment cannot be deleted.
4. Test the https://github.com/mautic/mautic/pull/11276 feature that it works with both membership filters.
5. CLOUD ONLY: Test that changing adding (or removing) a contact from a segment that is used in some segment membership filter will also rebuild the parent segments for that contact.

#### Other areas of Mautic that may be affected by the change:
1. The circular dependency validation based on the segment membership filter.

#### List deprecations along with the new alternative:
1. The method `isContactSegmentReference` was deprecated as unused with no replacement

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. The new segment membership - static filter
2. The circular dependency validation
